packages/langchain/src/openai/util.ts

```diff
@@ -0,0 +1,212 @@
+  res: OpenAiChatCompletionOutput
+): ChatResult {
+  return {
+    generations: res.choices.map((choice: OpenAiChatCompletionChoice) => ({
+      text: choice.message.content || '',
+      message: new AIMessage({
+        content: choice.message.content || '',
+        additional_kwargs: {
+          finish_reason: choice.finish_reason,
+          index: choice.index,
+          function_call: choice.message.function_call,
+          tool_calls: choice.message.tool_calls,
+          tool_call_id: ''
+        }
+      }),
+      generationInfo: {
+        finish_reason: choice.finish_reason,
+        index: choice.index,
+        function_call: choice.message.function_call,
+        tool_calls: choice.message.tool_calls
+      }
+    })),
+    llmOutput: {
+      created: res.created,
+      id: res.id,
+      model: res.model,
+      object: res.object,
+      tokenUsage: {
+        completionTokens: res.usage.completion_tokens,
+        promptTokens: res.usage.prompt_tokens,
+        totalTokens: res.usage.total_tokens
+      }
+    }
+  };
+}
+
+/**
+ * Maps {@link BaseMessage} to OpenAI Messages.
+ * @param message - The message to map.
+ * @returns The OpenAI Chat Message.
+ * @internal
+ */
+export function mapBaseMessageToOpenAiChatMessage(
+  message: BaseMessage
+): OpenAiChatMessage {
+  return {
+    content: message.content,
+    name: message.name,
+    role: mapBaseMessageToRole(message),
+    function_call: message.additional_kwargs.function_call,
+    tool_calls: message.additional_kwargs.tool_calls,
+    tool_call_id:
+      message._getType() === 'tool' ? (message as ToolMessage).tool_call_id : ''
+  } as OpenAiChatMessage;
+}
+
+/**
+ * Checks if a given array is a structured tool array.
+ * @param tools - The array to check.
+ * @returns Whether the array is a structured tool array.
+ * @internal
+ */
+export function isStructuredToolArray(
+  tools?: unknown[]
+): tools is StructuredTool[] {
+  return (
+    tools !== undefined &&
+    tools.every(tool => Array.isArray((tool as StructuredTool).lc_namespace))
+  );
+}
+
+/**
+ * Chunk an array into smaller arrays of specified chunk size.
+ * @param arr - Input array to be chunked.
+ * @param chunkSize - Size of each chunk.
+ * @returns Array of chunks.
+ * @internal
+ */
+export function chunkArray(arr: string[], chunkSize: number): string[][] {
+  return arr.reduce((chunks, elem, index) => {
+    const chunkIndex = Math.floor(index / chunkSize);
+    const chunk = chunks[chunkIndex] || [];
+
+    chunks[chunkIndex] = chunk.concat([elem]);
+    return chunks;
+  }, [] as string[][]);
+}
+
+/**
+ * Maps the langchain's input interface to our own client's input interface
+ * @param client The Langchain OpenAI client
+ * @param options The Langchain call options
+ * @param messages The messages to be send
+ * @returns A AI SDK compatibile request
+ * @internal
+ */
+export function mapLangchainToAiClient(
+  client: OpenAiChatClient,
+  options: OpenAiChatCallOptions,
```

Comment:
[q] Is this type really correct? I see `OpenAiChatCallOptions` is defined as:

```ts
export interface OpenAiChatCallOptions
  extends Omit<
      ChatOpenAICallOptions,
      | 'tool_choice'
      | 'promptIndex'
      | 'functions'
      | 'function_call'
      | 'tools'
      | 'response_format'
    >,
    Pick<
      OpenAiChatCompletionParameters,
      'tool_choice' | 'functions' | 'tools' | 'response_format'
    > {}
```

It omits `functions` and `tools` (which has `schema` property supporting zod schema) from langchain and pick them from our sdk, which uses json schema. But I assume, options here should still be in the original langchain format or?

Reply:
Realizing such type issues had to take a pretty in-depth review, so thanks for that already :D 
I think using our types here was correct since the same type was used to invoke any method of our langchain client.

I did change almost all types & classes by now though, so it should be (hopefully) clear which type we are intended to use.